### PR TITLE
Directly create a dylib device in the simple_embedding test

### PR DESCRIPTION
**Pull request by marbre**
Directly create a dylib device instead of leaving it to the driver.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]
20KB is quite a lot - and 1.1MB is way too much - would be good to understand what you're measuring :)
We still have some cmake work to do to get the build settings tuned but with `IREE_SIZE_OPTIMIZED=1` in a MinSizeRel build on MSVC, a few flags, and a few tweaks I can get decent numbers (and this is still a 64bit build - a 32bit build should be smaller):
```
      /DIREE_STATUS_MODE=0
      /DIREE_FLAGS_ENABLE_CLI=0
      /DIREE_VM_EXT_I64_ENABLE=0
      /DIREE_VM_EXT_F32_ENABLE=0
```

For the task-system based local driver I get 90KB:
image

And for the current smallest build with just the synchronous HAL + embedded elf loader I get 62KB:
image
Which then compresses down to 30KB (or 35,930B on disk with all alignment/PE headers, etc), so there's not a lot of entropy in there and it could probably be trimmed with better linker flags. It's still a lot bigger than I would like - in particular I think I can trim down the hal_module.c more (1.5-2KB of that is just the names of the exported functions used for linking).

The total code size contribution of the driver registry and the a driver is ~1KB.
So, we should probably figure out how to get a repeatable size-optimized build together before we make too many more decisions based on sizes that are a few orders of magnitude off :)

This PR is fine for this embedding sample as we can treat it as a whatever-is-needed-to-unblock-bare-metal-work. One thing I'd like to get setup on @antiagainst's new dashboards is a size benchmark so we can start to track this over time.

I'll try to get some of the settings I'm using checked in but haven't spent any time fiddling with size-optimized flags for clang/ld - but that would likely also differ for your target platform/toolchain so I'm not sure if the right place to do that stuff. Open to ideas!

File: iree/samples/simple_embedding/device_dylib.c [hunk 1]
```diff
@@ -16,20 +16,36 @@
...
-      iree_hal_driver_registry_default()));
-  // Create the hal driver from the name.
-  iree_hal_driver_t* driver = NULL;
+  // Set paramters for the device created in the next step.
+  iree_hal_task_device_params_t params;
+  iree_hal_task_device_params_initialize(&params);
+
+  iree_hal_executable_loader_t* loaders[2] = {NULL, NULL};
+  iree_host_size_t loader_count = 0;
+  (iree_hal_embedded_library_loader_create(iree_allocator_system(),
```
[the remark this answers is not in the document]
Comment (marbre) [reply iree/samples/simple_embedding/device_dylib.c:32]: Argh, you might have noticed that is already enclosed by `(  );`. I missed to re-add the `IREE_RETURN_IF_ERROR` macro. Thanks for the catch.

**Comment by marbre** [comment]
> 20KB is quite a lot - and 1.1MB is way too much - would be good to understand what you're measuring :)

I just checked ` ls -l --block-size=k simple_embedding_dylib` when building with `-DCMAKE_BUILD_TYPE` set to `Release`. So no real sophisticated analysis till now.

**Comment by benvanik** [comment]
> `-DCMAKE_BUILD_TYPE` set to `Release`. So no real sophisticated analysis till now.

Ah you want MinSizeRel - otherwise you are measuring DWARF symbols :)

**Comment by marbre** [comment]
> > `-DCMAKE_BUILD_TYPE` set to `Release`. So no real sophisticated analysis till now.
> 
> Ah you want MinSizeRel - otherwise you are measuring DWARF symbols :)

True! My CMake variants were messed up and it was still a Debug build. Actually, the change saves us 6K. It reduces the binary size from 323K to 317K.

**Pull request merged**